core/src/test/java/org/apache/iceberg/catalog/CatalogTransactionTests.java

```diff
@@ -0,0 +1,736 @@
+    } else {
+      Assertions.assertThatThrownBy(catalogTransaction::commitTransaction)
+          .isInstanceOf(CommitFailedException.class)
+          .hasMessageContaining("Requirement failed: current schema changed: expected id 0 != 1");
+    }
+
+    assertThat(baseMetadataOne).isSameAs(((BaseTable) one).operations().refresh());
+    assertThat(baseMetadataTwo).isSameAs(((BaseTable) two).operations().refresh());
+    assertThat(baseMetadataThree).isNotSameAs(((BaseTable) three).operations().refresh());
+    assertThat(((BaseTable) three).operations().refresh().schema().findField("new-column"))
+        .isNull();
+    assertThat(((BaseTable) three).operations().refresh().schema().findField("data")).isNull();
+    assertThat(((BaseTable) three).operations().refresh().schema().columns()).hasSize(1);
+  }
+
+  @Test
+  public void txAgainstMultipleTablesLastOneFails() {
+    txAgainstMultipleTablesLastOneFails(SNAPSHOT);
+  }
+
+  @Test
+  public void txAgainstMultipleTablesLastOneFailsWithSerializable() {
+    txAgainstMultipleTablesLastOneFails(SERIALIZABLE);
+  }
+
+  private void txAgainstMultipleTablesLastOneFails(
+      CatalogTransaction.IsolationLevel isolationLevel) {
+    for (String tbl : Arrays.asList("a", "b", "c")) {
+      catalog().createTable(TableIdentifier.of("ns", tbl), SCHEMA);
+    }
+
+    TableIdentifier first = TableIdentifier.of("ns", "a");
+    TableIdentifier second = TableIdentifier.of("ns", "b");
+    TableIdentifier third = TableIdentifier.of("ns", "c");
+    Table one = catalog().loadTable(first);
+    Table two = catalog().loadTable(second);
+    Table three = catalog().loadTable(third);
+
+    TableMetadata baseMetadataOne = ((BaseTable) one).operations().current();
+    TableMetadata baseMetadataTwo = ((BaseTable) two).operations().current();
+    TableMetadata baseMetadataThree = ((BaseTable) three).operations().current();
+
+    CatalogTransaction catalogTransaction = catalog().createTransaction(isolationLevel);
+    Catalog txCatalog = catalogTransaction.asCatalog();
+    txCatalog.loadTable(first).newAppend().appendFile(FILE_A).appendFile(FILE_B).commit();
+    assertThat(baseMetadataOne).isSameAs(((BaseTable) one).operations().refresh());
+
+    txCatalog.loadTable(second).newFastAppend().appendFile(FILE_B).appendFile(FILE_C).commit();
+    txCatalog.loadTable(second).newDelete().deleteFile(FILE_C).commit();
+    assertThat(baseMetadataTwo).isSameAs(((BaseTable) two).operations().refresh());
+
+    txCatalog.loadTable(third).newDelete().deleteFile(FILE_A).commit();
+    txCatalog.loadTable(third).newAppend().appendFile(FILE_D).commit();
+
+    assertThat(baseMetadataThree).isSameAs(((BaseTable) three).operations().refresh());
+
+    // perform updates outside the catalog TX
+    three.newAppend().appendFile(FILE_A).appendFile(FILE_D).commit();
+    Snapshot snapshot = ((BaseTable) three).operations().refresh().currentSnapshot();
+    assertThat(snapshot).isNotNull();
```

Comment:
I think you can assume that normal operations that do not go through `CatalogTransaction` behave the way you expect. That will allow you to avoid longer tests by omitting those assertions. It shouldn't be necessary to assert that the latest snapshot is non-null, nor should it be necessary to check that the `data` column was deleted in the example above.